best-practices/sdk-principles.md

```diff
@@ -0,0 +1,226 @@
+space.members.subscribe('leave', (member) => { ... })
+/* Shorthand for space.members.enter */
+space.enter({ 
+  username: "Claire Lemons",
+  avatar: "https://slides-internal.com/users/clemons.png",
+});
+space.channel // -> RealtimeChannel from Core
+space.channel.on('attached', () => { ... });
+spaces.connection // -> RealtimeConnection from Core
+spaces.connection.once('connected', () => { ... });
+spaces.client // -> Realtime from Core 
+```
+
+### Asset Tracking
+
+The AAT APIs are unlikely to be changed unless we decide to further develop the Asset Tracking product, however should it be developer further, we’d like to make it consistent with other SDKs as follows:
+
+```typescript
+import { Realtime } from '@ably/core';
+import { SubscriberClient } from '@ably/asset-tracking';
+const client = new Realtime(ABLY_API_KEY);
+const subClient = new SubscriberClient(client);
+// Get an asset.
+const asset = subClient.assets.get('my_tracking_id');
+asset.channel // -> RealtimeChannel from Core
+subClient.connection // -> RealtimeConnection from Core
+subClient.client // -> Realtime from Core 
+```
+
+### Notifications
+
+In the future, as we develop a more complete notification offering (not just push notifications, but consider a notification center where we provide APIs to handle delivering in-app notifications, @ mentions, manage read status, smart notification of devices only when a user has not seen it on their desktop etc), we may consider removing notifications from the current core SDKs into their own client, example:
+
+```typescript
+import { Realtime } from '@ably/core';
+import { PushNotificationsClient } from '@ably/notifications';
+import { PushAdminClient } from '@ably/notifications';
+const client = new Realtime(ABLY_API_KEY);
+const pushClient = new PushNotificationsClient(client);
+pushClient.activate();
+const pushChannel = pushClient.channels.get('foo');
+pushchannel.channel // -> RealtimeChannel from Core
+pushClient.connection // -> RealtimeConnection from Core
+pushClient.client // -> Realtime from Core 
+```
+
+## Best practice guidelines for SDKs
+
+In no particular order, below is a list of best practices and guidelines for each SDK. Whilst the right architecture, naming and structure for each SDK will be decided by each team working on this software, we need to be intentional and have a solid basis when we deviating from broader best practice, conventions and learnings we have from building SDKs over the last 7 years.
+
+> It’s an anti-pattern to think that any of the following points can be deferred until the first Release Candidate or General Availability release. Ignoring these best practices from the get go will likely lead to bugs, quality issues experience by customers, and delays when trying to move a release into RC/GA state. A move to RC/GA should not typically require a refactor / considerable work, it reaches that state because the software has proven to be stable enough to mature one of these states 
```

Comment:
```suggestion
> [!WARNING]
> It’s an anti-pattern to think that any of the following points can be deferred until the first Release Candidate or General Availability release. Ignoring these best practices from the get go will likely lead to bugs, quality issues experience by customers, and delays when trying to move a release into RC/GA state. A move to RC/GA should not typically require a refactor / considerable work, it reaches that state because the software has proven to be stable enough to mature one of these states 
```

original doc has this as a warning

Reply:
thanks, fixed